### ISSUE TITLE
map matching score using time values

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#620]
so we recently added an `epoch_time` value to the `Measurement` and `MatchResult` objects in `meili`. i was wondering if when map matching we could make use of those to down weight paths between certain candidates. basically my thought is that if a certain candidate route between points has a computed time of `X` but the difference in `epoch_time` on the `Measurements` is `Y` and they are very far apart from each other (especially if `X` is way longer than `Y`), then we should say that this path is not likely (or maybe invalid).

@ptpt am i making any sense on this? could you give me some pointers on where i might change the code to affect the weights based on the difference/similarity of the time values? mainly if someone gives me a trace that seems to have problems somewhere in it, i'd prefer that we didnt find a path between some of the pairs of the points. if this isnt making sense let me know and i can draw a picture to illustrate it.

**Comment by ptpt** [comment]
@kevinkreiser Yes, totally make sense.

Two places need to be changed to support weighting based on travel time.

1. put travel time into Label https://github.com/valhalla/valhalla/blob/master/src/meili/routing.cc#L507-L513 (This part is tricky as the code is getting messy there. I'd be happy to assign myself to add it).

2. calculate transition cost based on travel time https://github.com/valhalla/valhalla/blob/master/src/meili/map_matching.cc#L159-L202

```c++
float
MapMatching::TransitionCost(const State& left, const State& right) const
{
  if (!left.routed()) {
// lots of code omitted
  }

  const auto label = left.last_label(right);
  if (label) {
    // travel time from left candidate to right candidate
    const auto X = label->travel_time;
    // time diff between left measurement to right measurement
    const auto Y = measurement(right).epoch_time - measurement(left).epoch_time;
    // calculate transition cost based on above
    return CalculateTransitionCost(label->turn_cost, X, Y);
  }

  return -1.f;
}
```

**Comment by kevinkreiser** [comment]
thanks @ptpt any help here would be greatly appreciated! i was wondering if you thought it would be a good idea. specifically i was also wondering that if all of the potential routes between `left` and `right` states seem to have very low probability due to the time difference being so high, will the algorithm assume that none of them are valid or will it still pick the highest probability one?

the specific problem im running into is a gps trace looks pretty good but then has a horrible spot in it where it just makes no sense. in this case i would prefer to get two disconnected map matched results and just let the horrible spot un-matched. would this be useful to mapillary as well? the idea is that no data for a given set of bad points is better than wrong data :smile:

**Comment by ptpt** [comment]
> I was wondering if you thought it would be a good idea.

Yes it is. I believe that using time difference would improve the matching accuracy.

> specifically i was also wondering that if all of the potential routes between left and right states seem to have very low probability due to the time difference being so high, will the algorithm assume that none of them are valid or will it still pick the highest probability one?

No, the algorithm won't pick the highest one if two measurements are too far apart (further than `breakage_distance `), or time difference in terms of travel time is too high. Instead it will assume that the two are disconnected. However it doesn't mean two measurements can't find good matches. It's still possible to find paths from previous measurement to the left measurement, or from the right measurement to its next measurement.

> the specific problem im running into is a gps trace looks pretty good but then has a horrible spot in it where it just makes no sense. in this case i would prefer to get two disconnected map matched results and just let the horrible spot un-matched.

Yeah I think `max_route_distance_factor` is exactly designed to solve this problem. We can make it work for the time-based weighting as well, but with a different name, e.g. `max_travel_time_factor`. If from left states to right states we can not find a path that uses less than `max_travel_time_factor * (right measurement epoch_time  - left measurement epoch_time)`, then we mark the two measurements as disconnected.

> would this be useful to mapillary as well? the idea is that no data for a given set of bad points is better than wrong data 😄

Yes! would be very useful as all images have capture times.

**Comment by ptpt** [comment]
Assigned to me. Will give it a try this weekend.

**Comment by kevinkreiser** [comment]
@ptpt thats awesome i'm glad it will benefit us both! sorry it took me so long to get up to speed on map matching otherwise we could have gotten to this sooner!

one thing i should mention is that when there isnt a time on a given measurement it will have `-1` as the value. so if either `left measurement epoch_time` or `right measurement epoch_time` is `-1` then we'll want to avoid checking that the path time is within range.

**Comment by ptpt** [comment]
Great! If time is not available, then we can fallback to the route distance based weighting.

**Comment by kevinkreiser** [comment]
ok ive taken a stab at this and ran into a few questions!

1. when deciding whether or not to put a label in the queue, we check the sortcost instead of the cost, this seems to be the mechanism by which we avoid allowing routes greater than the `max_route_distance_factor` or the `breakage_distance` (whichever is smaller). it looks like this:

```c++
 if (sortcost < max_cost_) {
```

I was thinking it needed to be changed to:

```c++
if (cost.cost < max_dist_ && (max_time_ == 0.f || cost.secs < max_time_)) {
```

note that i changed `max_cost_` to `max_dist_` and also note that `max_time_` being `0` means that there were no timestamps on the measurements. @ptpt what do you think is it wrong to change the distance comparison from `sortcost` to `cost.cost`?

2. i'm not sure how i should change `CalculateTransitionCost`. basically the units seem to be a mixture of time and distance, is it ok to just add in the difference in time? currently it looks like this:

```c++
{ return (turncost + std::abs(route_distance - measurement_distance)) * inv_beta_; }
```

can i just change it to:

```c++
{ return (turncost + std::abs(route_distance - measurement_distance) + std::abs(route_time - measurement_distance)) * inv_beta_; }
```

**Comment by ptpt** [comment]
1. It should be fine since we have switched to `baldr::DoubleBucketQueue` which can hold greater sort costs than the specified range.

2. It doesn't make sense to combine different metrics because the time could be much larger than the distance (e.g. milisecond and meters). This would make weights unbalance. It would be better to combine them as a single speed difference:

```c++
auto expected_speed = route_distance / route_time;
auto actual_speed = measurement_distance / measurement_time;

return std::abs(actual_speed - expected_speed) * inv_beta_;
```

Note that I removed the turn cost because I assumed that its impact is reflected in `route_time` (sif has considered turns)
